validator/client/validator.go

```diff
@@ -135,53 +142,74 @@ func (v *validator) SlotDeadline(slot uint64) time.Time {
 		// Do nothing if not epoch start AND assignments already exist.
 		return nil
 	}
 
 	ctx, span := trace.StartSpan(ctx, "validator.UpdateAssignments")
 	defer span.End()
-
+	pks := [][]byte{}
+	for i := range v.keys {
+		pks = append(pks, v.keys[i].PublicKey.Marshal())
+	}
 	req := &pb.CommitteeAssignmentsRequest{
 		EpochStart: slot,
-		PublicKey:  [][]byte{v.key.PublicKey.Marshal()},
+		PublicKey:  pks,
 	}
 
 	resp, err := v.validatorClient.CommitteeAssignment(ctx, req)
 	if err != nil {
-		v.assignment = nil // Clear assignments so we know to retry the request.
+		v.assignments = nil // Clear assignments so we know to retry the request.
 		return err
 	}
 
-	v.assignment = resp
+	v.assignments = resp
+	for _, assignment := range v.assignments.Assignment {
+		var proposerSlot uint64
+		var attesterSlot uint64
+		if assignment.IsProposer {
+			proposerSlot = assignment.Slot
+			attesterSlot = assignment.Slot
+		} else {
+			attesterSlot = assignment.Slot
+		}
 
-	lFields := logrus.Fields{
-		"attesterSlot": resp.Assignment[0].Slot - params.BeaconConfig().GenesisSlot,
-		"proposerSlot": "Not proposing",
-		"shard":        resp.Assignment[0].Shard,
-	}
-	if v.assignment.Assignment[0].IsProposer {
-		lFields["proposerSlot"] = resp.Assignment[0].Slot - params.BeaconConfig().GenesisSlot
-	}
+		lFields := logrus.Fields{
+			"attesterSlot": attesterSlot - params.BeaconConfig().GenesisSlot,
+			"proposerSlot": "Not proposing",
+			"shard":        assignment.Shard,
+		}
+		if assignment.IsProposer {
+			lFields["proposerSlot"] = proposerSlot - params.BeaconConfig().GenesisSlot
+		}
+
+		log.WithFields(lFields).Info("Updated validator assignments")
 
-	log.WithFields(lFields).Info("Updated validator assignments")
+	}
 	return nil
 }
 
-// RoleAt slot returns the validator role at the given slot. Returns nil if the
-// validator is known to not have a role at the at slot. Returns UNKNOWN if the
+// RolesAt slot returns the validator roles at the given slot. Returns nil if the
+// validator is known to not have a roles at the at slot. Returns UNKNOWN if the
 // validator assignments are unknown. Otherwise returns a valid ValidatorRole.
-func (v *validator) RoleAt(slot uint64) pb.ValidatorRole {
-	if v.assignment == nil {
-		return pb.ValidatorRole_UNKNOWN
-	}
-	if v.assignment.Assignment[0].Slot == slot {
-		if v.assignment.Assignment[0].IsProposer {
+func (v *validator) RolesAt(slot uint64) map[string]pb.ValidatorRole {
+	rolesAt := make(map[string]pb.ValidatorRole)
+	for _, assignment := range v.assignments.Assignment {
+		var role pb.ValidatorRole
+		if assignment == nil {
+			role = pb.ValidatorRole_UNKNOWN
+		}
+		if assignment.Slot == slot {
 			// Note: A proposer also attests to the slot.
-			return pb.ValidatorRole_PROPOSER
+			if assignment.IsProposer {
+				role = pb.ValidatorRole_PROPOSER
+			} else {
+				role = pb.ValidatorRole_ATTESTER
+			}
 		}
-		return pb.ValidatorRole_ATTESTER
+		role = pb.ValidatorRole_UNKNOWN
```

Comment:
This sets the role to be UNKNOWN every time

Reply:
added it in else